### PR TITLE
[FEATURE] Agent split

### PR DESCRIPTION
**Pull request by deveusss**
to be done:
- [x] Agents Split
- [x] Tests
- [ ] Comments
- [x] Typescript types
- [ ] Readme/docs

### CODE REVIEW

**Comment by deveusss** [comment]
to be done:
[] Tests
[] Comments
[] Typescript stuff

File: contracts/croncat-agents/src/contract.rs [hunk 1]
```diff
@@ -1,31 +1,538 @@
...
 #[cfg(not(feature = "library"))]
 use cosmwasm_std::entry_point;
-use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
+use cosmwasm_std::{
+    has_coins, to_binary, Addr, BankMsg, Binary, Coin, Deps, DepsMut, Env, MessageInfo, Response,
+    StdError, StdResult, Storage, SubMsg, Uint128,
+};
+use croncat_sdk_agents::msg::{AgentResponse, AgentTaskResponse, GetAgentIdsResponse};
+use croncat_sdk_agents::types::{Agent, AgentStatus};
+use croncat_sdk_core::types::Config;
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:24]:
```suggestion
use croncat_sdk_agents::types::Config;
```
Agents should have it's own config!

File: contracts/croncat-agents/src/contract.rs [hunk 2]
```diff
@@ -1,31 +1,538 @@
...
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn instantiate(
-    mut _deps: DepsMut,
+    deps: DepsMut,
     _env: Env,
     _info: MessageInfo,
     _msg: InstantiateMsg,
 ) -> Result<Response, ContractError> {
-    todo!();
+    let config: Config = CONFIG.load(deps.storage)?;
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:36]: Where does it load CONFIG from? :rofl:

File: packages/croncat-sdk-agents/src/msg.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        cost: u128,
+    },
+    UpdateAgent {
+        payable_account_id: String,
+    },
+    CheckInAgent {},
+    UnregisterAgent {
+        from_behind: Option<bool>,
+    },
+    WithdrawReward {},
```
Comment (Buckram123) [inline packages/croncat-sdk-agents/src/msg.rs:22]:
```suggestion
    WithdrawReward {},
    NewTaskCreated {},
```
Can you implement please this logic: 
https://github.com/CronCats/cw-croncat/blob/b920d1a5d5b1730397b04acdbba205eacfd487bd/contracts/cw-croncat/src/tasks.rs#L380-L395

File: packages/croncat-sdk-agents/src/msg.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+#[cw_serde]
+pub enum AgentQueryMsg {
```
Comment (Buckram123) [inline packages/croncat-sdk-agents/src/msg.rs:26]:
```suggestion
#[cw_serde]
#[derive(QueryResponses)]
pub enum AgentQueryMsg {
```
Let's use `QueryResponse` for easier schema handle!

File: packages/croncat-sdk-agents/src/msg.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+pub struct GetAgentIdsResponse {
+    pub active: Vec<Addr>,
+    pub pending: Vec<Addr>,
+}
+
+#[cw_serde]
+pub struct AgentResponse {
+    pub status: AgentStatus,
+    pub payable_account_id: Addr,
+    pub balance: Uint128,
```
Comment (Buckram123) [inline packages/croncat-sdk-agents/src/msg.rs:48]: Do we want to return `rewards_balance` here or only from manager contract? @TrevorJTClarke

File: packages/croncat-sdk-agents/src/msg.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    pub register_start: Timestamp,
+}
+
+#[cw_serde]
+pub struct AgentTaskResponse {
+    pub num_block_tasks: Uint64,
+    pub num_block_tasks_extra: Uint64,
+    pub num_cron_tasks: Uint64,
+    pub num_cron_tasks_extra: Uint64,
+}
```
Comment (Buckram123) [inline packages/croncat-sdk-agents/src/msg.rs:60]: Should think of better response for querying with offset!

File: packages/croncat-sdk-core/Cargo.toml [hunk 7]
```diff
@@ -5,4 +5,8 @@ authors = []
 edition = "2021"
 repository = "https://github.com/CronCats/cw-croncat"
 
-[dependencies]
\ No newline at end of file
+[dependencies]
+cosmwasm-std = { workspace = true }
+cosmwasm-schema = { workspace = true }
+thiserror = { workspace = true }
+cw20 = { workspace = true }
```
Comment (Buckram123) [inline packages/croncat-sdk-core/Cargo.toml:12]: I don't think `packages/croncat-sdk-core/*` should be used for agents yet, lets revert changes here!

File: packages/croncat-sdk-tasks/Cargo.toml [hunk 8]
```diff
@@ -5,4 +5,7 @@ authors = []
 edition = "2021"
 repository = "https://github.com/CronCats/cw-croncat"
 
-[dependencies]
\ No newline at end of file
+[dependencies]
+cosmwasm-std = { workspace = true }
+cosmwasm-schema = { workspace = true }
+croncat-sdk-core = { workspace = true }
```
Comment (Buckram123) [inline packages/croncat-sdk-tasks/Cargo.toml:11]: revert it too!

File: packages/croncat-sdk-tasks/src/types.rs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    Block,
+    Cron,
+}
+#[cw_serde]
+pub struct TaskInfo {
+    pub task_hash: Vec<u8>,
+    pub task_is_extra: Option<bool>,
+    pub agent_id: Addr,
+    pub slot_kind: SlotType,
+}
```
Comment (Buckram123) [inline packages/croncat-sdk-tasks/src/types.rs:15]: Is it needed for agents contract?

File: contracts/croncat-agents/src/contract.rs [hunk 10]
```diff
@@ -1,31 +1,538 @@
...
+            payable_account_id,
+            cost,
+        } => register_agent(deps, info, env, payable_account_id, cost),
+        ExecuteMsg::UpdateAgent { payable_account_id } => {
+            update_agent(deps, info, env, payable_account_id)
+        }
+        ExecuteMsg::UnregisterAgent { from_behind } => {
+            unregister_agent(deps.storage, &info.sender, from_behind)
+        }
+        ExecuteMsg::WithdrawReward {} => withdraw_agent_balance(deps, &info.sender),
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:77]:
```suggestion
```
This method should be only on manager contract, I believe!

File: contracts/croncat-agents/src/contract.rs [hunk 11]
```diff
@@ -1,31 +1,538 @@
...
+        a
+    } else {
+        return Ok(None);
+    };
+
+    let agent_status = get_agent_status(deps.storage, env, account_id, total_tasks)
+        // Return wrapped error if there was a problem
+        .map_err(|err| StdError::GenericErr {
+            msg: err.to_string(),
+        })?;
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:100]: I would like to have typed errors, so it's easier to test and debug!

File: contracts/croncat-agents/src/contract.rs [hunk 12]
```diff
@@ -1,31 +1,538 @@
...
+    Ok(Some(agent_response))
+}
+
+/// Get a list of agent addresses
+fn query_get_agent_ids(deps: Deps) -> StdResult<GetAgentIdsResponse> {
+    let active: Vec<Addr> = ACTIVE_AGENTS.load(deps.storage)?;
+    let pending: Vec<Addr> = PENDING_AGENTS
+        .iter(deps.storage)?
+        .collect::<StdResult<Vec<Addr>>>()?;
+
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:119]: can we get pagination here?

File: contracts/croncat-agents/src/contract.rs [hunk 13]
```diff
@@ -1,31 +1,538 @@
...
+    deps: Deps,
+    env: Env,
+    account_id: String,
+    slots: (u64, u64), //block_slots,cron_slots
+) -> StdResult<Option<AgentTaskResponse>> {
+    let account_id = deps.api.addr_validate(&account_id)?;
+    let active = ACTIVE_AGENTS.load(deps.storage)?;
+    if !active.contains(&account_id) {
+        return Err(StdError::GenericErr {
+            msg: "Agent is not active!".to_owned(),
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:133]: let's add typed errors!

File: contracts/croncat-agents/src/contract.rs [hunk 14]
```diff
@@ -1,31 +1,538 @@
...
+fn register_agent(
+    deps: DepsMut,
+    info: MessageInfo,
+    env: Env,
+    payable_account_id: Option<String>,
+    cost: u128,
+) -> Result<Response, ContractError> {
+    if !info.funds.is_empty() {
+        return Err(ContractError::InsufficientFunds);
+    }
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:159]: What do you mean by `InsufficientFunds` :smile:
It should return "Doesn't accept deposit" or something like that

File: contracts/croncat-agents/src/contract.rs [hunk 15]
```diff
@@ -1,31 +1,538 @@
...
+    let account = info.sender;
+
+    // REF: https://github.com/CosmWasm/cw-tokens/tree/main/contracts/cw20-escrow
+    // Check if native token balance is sufficient for a few txns, in this case 4 txns
+    let agent_wallet_balances = deps.querier.query_all_balances(account.clone())?;
+    if !has_coins(&agent_wallet_balances, &Coin::new(cost, c.native_denom))
+        || agent_wallet_balances.is_empty()
+    {
+        return Err(ContractError::InsufficientFunds);
+    }
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:174]: Do we need this logic at all? And what's our flow for that @TrevorJTClarke?

File: contracts/croncat-agents/src/contract.rs [hunk 16]
```diff
@@ -1,31 +1,538 @@
...
+        vec![SubMsg::new(BankMsg::Send {
+            to_address: to.into(),
+            amount: vec![Coin {
+                denom: config.native_denom,
+                amount: balance,
+            }],
+        })]
+    };
+    Ok((msgs, balance))
+}
```
Comment (Buckram123) [inline contracts/croncat-agents/src/contract.rs:289]: It shouldn't send coins ever

File: contracts/croncat-agents/src/contract.rs (hunk 1 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:24]: yes, but from this config I get native_denom

File: contracts/croncat-agents/src/contract.rs (hunk 16 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:289]: what do u mean, more details please @Buckram123

File: contracts/croncat-agents/src/contract.rs (hunk 1 above)
Comment (Buckram123) [reply contracts/croncat-agents/src/contract.rs:24]: > yes, but from this config I get native_denom

Why do you need native_denom?

File: contracts/croncat-agents/src/contract.rs (hunk 16 above)
Comment (Buckram123) [reply contracts/croncat-agents/src/contract.rs:289]: > what do u mean, more details please @Buckram123

Coins should not leave manager!

Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:289]: but coins here is just agent balance going to reward address, are we moving rewards part into manager?

File: contracts/croncat-agents/src/contract.rs (hunk 2 above)
Comment (mikedotexe) [reply contracts/croncat-agents/src/contract.rs:36]: Yeah I think this is giving me an error

File: contracts/croncat-agents/src/contract.rs (hunk 1 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:24]: for sending rewards

File: packages/croncat-sdk-tasks/Cargo.toml (hunk 8 above)
Comment (deveusss) [reply packages/croncat-sdk-tasks/Cargo.toml:11]: we have helper stuff in croncat-sdk-core

File: packages/croncat-sdk-agents/src/msg.rs (hunk 6 above)
Comment (deveusss) [reply packages/croncat-sdk-agents/src/msg.rs:68]: yeah refactoring is needed for the whole project))

File: packages/croncat-sdk-core/Cargo.toml (hunk 7 above)
Comment (deveusss) [reply packages/croncat-sdk-core/Cargo.toml:12]: why? croncat-sdk-core can contain helper methods

File: packages/croncat-sdk-tasks/src/types.rs (hunk 9 above)
Comment (deveusss) [reply packages/croncat-sdk-tasks/src/types.rs:15]: yes in balancer

File: contracts/croncat-agents/src/contract.rs [hunk 17]
```diff
@@ -1,31 +1,548 @@
...
+            payable_account_id,
+            cost,
+        } => register_agent(deps, info, env, payable_account_id, cost),
+        ExecuteMsg::UpdateAgent { payable_account_id } => {
+            update_agent(deps, info, env, payable_account_id)
+        }
+        ExecuteMsg::UnregisterAgent { from_behind } => {
+            unregister_agent(deps.storage, &info.sender, from_behind)
+        }
+        ExecuteMsg::CheckInAgent {} => accept_nomination_agent(deps, info, env),
```
Comment (mikedotexe) [inline contracts/croncat-agents/src/contract.rs:88]: I noticed that in commit fd670e9520130a068c42f995b6e3d6324c972b9b, the `WithdrawReward` variant was removed? I think that's causing `just optimize` to fail with:

```
error[E0004]: non-exhaustive patterns: `ExecuteMsg::WithdrawReward {  }` not covered
```

I reverted that commit (locally) and it's building now. Think we're going to need that back

**Review by mikedotexe — CHANGES_REQUESTED** [review]

File: packages/croncat-sdk-core/Cargo.toml (hunk 7 above)
Comment (Buckram123) [reply packages/croncat-sdk-core/Cargo.toml:12]: Isn't `croncat-sdk-core` for `croncat-manager` contract? Is there any helpers that will get used in every contract?

Comment (deveusss) [reply packages/croncat-sdk-core/Cargo.toml:12]: why not croncat-sdk-manager for croncat-manager? we should be consistent here!

File: contracts/croncat-agents/src/contract.rs (hunk 11 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:111]: Multiple errors could be caught here, we don't know the exact error

File: contracts/croncat-agents/src/contract.rs (hunk 10 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:88]: When agent is removed, we are calling withdraw rewards, how to deal with that if we are removing WithdrawReward? @Buckram123

Comment (Buckram123) [reply contracts/croncat-agents/src/contract.rs:88]: @deveusss should get called "remove_manager" on manager contract too!

File: contracts/croncat-agents/src/contract.rs (hunk 13 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:147]: √

File: packages/croncat-sdk-agents/src/msg.rs (hunk 3 above)
Comment (deveusss) [reply packages/croncat-sdk-agents/src/msg.rs:26]: √

File: contracts/croncat-agents/src/contract.rs (hunk 14 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs:173]: √

**Review by mikedotexe — CHANGES_REQUESTED** [review]
When I run `just optimize` I am getting this error:

```
error[E0433]: failed to resolve: could not find `macos` in `os`
 --> contracts/croncat-agents/src/balancer.rs:1:14
  |
1 | use std::os::macos::raw::stat;
  |              ^^^^^ could not find `macos` in `os`
```

File: contracts/croncat-agents/src/contract.rs (hunk 11 above)
Comment (Buckram123) [reply contracts/croncat-agents/src/contract.rs:122]:
```suggestion
?
```
Do we need to wrap it into generic then? @deveusss

File: contracts/croncat-agents/src/contract.rs (hunk 15 above)
Comment (mikedotexe) [reply contracts/croncat-agents/src/contract.rs]: I think so, as discussed in Discord. it's working now after recent commits. good eye tho

**Pull request merged**